### PR TITLE
Dsc_CMGroupDiscovery resource creation.

### PR DESCRIPTION
**Pull request by jeffotterpohl**
#### Pull Request (PR) description
Adding additional functionality to to being able to manage GroupDiscovery within the resource.

#### Task list

- [X] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [X] Resource documentation added/updated in README.md.
- [X] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [X] Comment-based help added/updated.
- [X] Localization strings added/updated in all localization files as appropriate.
- [X] Examples appropriately added/updated.
- [X] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [X] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/configmgrcbdsc/97)

### CODE REVIEW

File: source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1 [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+                $removing = @()
+                $missing = @()
+
+                foreach ($item in $compare)
+                {
+                    if ($item.SideIndicator -eq '<=')
+                    {
+                        if ($state.GroupDiscoveryScope.Name -contains $item.Name)
+                        {
+                            Write-Verbose -Message ($script:localizedData.GdsUpdate -f $item.Name, $item.LdapLocation, $item.Recurse,$item.LdapLocation, $item.Recurse)
```
Comment (NEllis280) [inline source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1:402]: Space after recurse, for consistency.

File: source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1 [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+                        }
+
+                        [array]$groupOus += New-CMADGroupDiscoveryScope @cmAddGroup
+                    }
+
+                    if ($GroupDiscoveryScope)
+                    {
+                        if ($item.SideIndicator -eq '=>')
+                        {
+                            #if ($refObject.Name -notcontains $item.Name)
```
Comment (NEllis280) [inline source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1:423]: Remove commented-out code.

File: source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1 [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+            $compare = Compare-Object @compareParams
+            $removing = @()
+            $missing = @()
+            foreach ($item in $compare)
+            {
+                if ($item.SideIndicator -eq '<=')
+                {
+                    if ($state.GroupDiscoveryScope.Name -contains $item.Name)
+                    {
+                        Write-Verbose -Message ($script:localizedData.GdsUpdate -f $item.Name, $item.LdapLocation, $item.Recurse,$item.LdapLocation, $item.Recurse)
```
Comment (NEllis280) [inline source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1:719]: add a space after recurse, for consistency

File: source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1 [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+                    }
+
+                    $result = $false
+                }
+
+                if ($GroupDiscoveryScope)
+                {
+                    if ($item.SideIndicator -eq '=>')
+                    {
+                        #if ($refObject.Name -notcontains $item.Name)
```
Comment (NEllis280) [inline source/DSCResources/DSC_CMGroupDiscovery/DSC_CMGroupDiscovery.psm1:733]: remove commented out code

File: source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1 [hunk 5]
```diff
@@ -4240,6 +4874,43 @@ function Set-ConfigMgrCBDscReverse
...
+            $wGrpDiscovery = "$resourceName = @{`r`n"
+
+            if ($groupDisc -eq 'INACTIVE')
+            {
+                $params = @{
+                    ResourceName = $resourceName
+                    SiteCode     = $SiteCode
+                    Indent       = 1
+                    Count        = 7
+                    #ExcludeList  = @('SiteCode','ScheduleInterval','ScheduleCount')
```
Comment (NEllis280) [inline source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1:4894]: remove commented out code

File: tests/Unit/CMGroupDiscovery.tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,919 @@
...
+                    IsGMT          = $false
+                    MinuteDuration = 0
+                    MinuteSpan     = 45
+                }
+
+                Mock -CommandName Import-ConfigMgrPowerShellModule
+                Mock -CommandName Set-Location
+            }
+
+            Context 'When retrieving Collection settings' {
```
Comment (NEllis280) [inline tests/Unit/CMGroupDiscovery.tests.ps1:225]: context should be 'When retrieving Group Discovery settings'

File: tests/Unit/CMGroupDiscovery.tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,919 @@
...
+                    $result.DayOfWeek                           | Should -Be -ExpectedValue 'Friday'
+                    $result.MonthlyWeekOrder                    | Should -Be -ExpectedValue 'First'
+                    $result.GroupDiscoveryScope                 | Should -BeOfType '[Microsoft.Management.Infrastructure.CimInstance]'
+                    $result.GroupDiscoveryScope[0].Name         | Should -Be -ExpectedValue 'Test1'
+                    $result.GroupDiscoveryScope[0].LdapLocation | Should -Be -ExpectedValue 'LDAP://OU=Test1,DC=contoso,DC=com'
+                    $result.GroupDiscoveryScope[0].Recurse      | Should -Be -ExpectedValue $true
+                    $result.GroupDiscoveryScope[1].Name         | Should -Be -ExpectedValue 'Test2'
+                    $result.GroupDiscoveryScope[1].LdapLocation | Should -Be -ExpectedValue 'LDAP://OU=Test2,DC=contoso,DC=com'
+                    $result.GroupDiscoveryScope[1].Recurse      | Should -Be -ExpectedValue $false
+                    #$result.GroupDiscoveryScope                | Should -Be -ExpectedValue $adContainersReturn
```
Comment (NEllis280) [inline tests/Unit/CMGroupDiscovery.tests.ps1:287]: remove commented out code

File: tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1 [hunk 8]
```diff
@@ -2641,6 +2769,96 @@ InModuleScope $script:subModuleName {
...
+                EnableFilteringExpiredPassword      = $true
+                GroupDiscoveryScope                 = @(
+                    @{
+                        Name           = 'Test1'
+                        LdapLocation   = 'OU=Test1,DC=contoso,DC=com'
+                        Recurse        = $true
+                        PSComputerName = 'localhost'
+                    }
+                )
+                GroupDiscoveryScopeToExclude        = $null
```
Comment (NEllis280) [inline tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1:2821]: spacing

File: tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1 [hunk 9]
```diff
@@ -3553,22 +3773,23 @@ InModuleScope $script:subModuleName {
...
                 Mock -CommandName Get-DscResource -MockWith { $getDscResourceReturn }
-                Mock -CommandName Invoke-DscResource -MockWith { $invokeCMAssetIntelligencePointNoneSchedule } -ParameterFilter { $Name -eq 'CMAssetIntelligencePoint' }
+                Mock -CommandName Invoke-DscResource -MockWith { $invokeForestDiscoveryDisabled }
+                Mock -CommandName Get-CMDiscoveryMethod -MockWith { $getForestDiscoveryDisabled } -ParameterFilter { $Name -eq 'ActiveDirectoryForestDiscovery' }
 
-                $result = Set-ConfigMgrCBDscReverse @assetIntell
+                $result = Set-ConfigMgrCBDscReverse @forestDiscovery
                 $result | Should -BeOfType System.String
-                $result | Should -Match 'CMAssetIntelligencePoint'
+                $result | Should -Match "CMForestDiscovery"
```
Comment (NEllis280) [inline tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1:3783]: Single quotes

File: README.md [hunk 10]
```diff
@@ -1472,8 +1530,7 @@ After importing the module, Set-ConfigMgrCBDscReverse will be available.
   ClientPush|ClientStatusSettings|CollectionEvaluationComponent|Collections|
   DistributionGroups|DistributionPoint|DistributionPointGroupMembers|EmailNotificationComponent|
   FallbackPoints|ForestDiscovery|HeartbeatDiscovery|MaintenanceWindow|ManagementPoint|
-  NetworkDiscovery|PullDistributionPoint|PxeDistributionPoint|
-  ReportingServicesPoint|SecurityScopes|ServiceConnection|SiteMaintenance|
+  NetworkDiscovery|PullDistributionPoint|PxeDistributionPoint|GroupDiscovery|  ReportingServicesPoint|SecurityScopes|ServiceConnection|SiteMaintenance|
```
Comment (NEllis280) [inline README.md:1533]: newline needed

File: tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1 (hunk 8 above)
Comment (jeffotterpohl) [reply tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1:2821]: This lines up on char 53 as above with the other items in the hash table

**Review by NEllis280 — APPROVED** [review]
Code Review complete. Tested code and additions work as expected.

**Pull request merged**